solidity/contracts/modules/MultisigModule.sol

```diff
@@ -0,0 +1,335 @@
+        EnumerableSet.AddressSet storage _validatorSet = validatorSet[_domain];
+        uint256 _validatorCount = _validatorSet.length();
+        address[] memory _validators = new address[](_validatorCount);
+        for (uint256 i = 0; i < _validatorCount; i++) {
+            _validators[i] = _validatorSet.at(i);
+        }
+        return _validators;
+    }
+
+    /**
+     * @notice Returns the number of validators enrolled in the validator set.
+     * @param _domain The remote domain of the validator set.
+     * @return The number of validators enrolled in the validator set.
+     */
+    function validatorCount(uint32 _domain) public view returns (uint256) {
+        return validatorSet[_domain].length();
+    }
+
+    // ============ Internal Functions ============
+
+    /**
+     * @notice Updates the commitment to the validator set for `_domain`.
+     * @param _domain The remote domain of the validator set.
+     * @return The commitment to the validator set for `_domain`.
+     */
+    function _updateCommitment(uint32 _domain) internal returns (bytes32) {
+        address[] memory _validators = validators(_domain);
+        uint256 _threshold = threshold[_domain];
+        bytes32 _commitment = keccak256(
+            abi.encodePacked(_threshold, _validators)
+        );
+        commitment[_domain] = _commitment;
+        return _commitment;
+    }
+
+    /**
+     * @notice Verifies the merkle proof of `_message` against the provided
+     * checkpoint.
+     * @param _metadata ABI encoded module metadata (see MultisigModuleMetadata.sol)
+     * @param _message Formatted Hyperlane message (see Message.sol).
+     */
+    function _verifyMerkleProof(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) internal pure returns (bool) {
+        // calculate the expected root based on the proof
+        bytes32 _calculatedRoot = MerkleLib.branchRoot(
+            _message.id(),
+            _metadata.proof(),
+            _message.nonce()
+        );
+        return _calculatedRoot == _metadata.root();
+    }
+
+    /**
+     * @notice Verifies that a quorum of the origin domain's validators signed
+     * the provided checkpoint.
+     * @param _metadata ABI encoded module metadata (see MultisigModuleMetadata.sol)
+     * @param _message Formatted Hyperlane message (see Message.sol).
+     */
+    function _verifyValidatorSignatures(
+        bytes calldata _metadata,
+        bytes calldata _message
+    ) internal view returns (bool) {
+        uint256 _threshold = _metadata.threshold();
+        bytes32 _digest;
+        {
+            uint32 _origin = _message.origin();
+            bytes32 _commitment = keccak256(
+                abi.encodePacked(_threshold, _metadata.validators())
+            );
+            // Ensures the validator set encoded in the metadata matches
+            // what we've stored on chain.
+            require(_commitment == commitment[_origin], "!commitment");
```

Comment:
possibly worth flagging that this line is what protects against someone being able to verify a message for an origin that hasn't been set because the `_commitment` above will always be non-zero, even if the metadata threshold and validators are non-zero